stage1/usr_from_kvm/lkvm/patches/dead_lock_fix.patch

```diff
@@ -0,0 +1,57 @@
+diff --git a/builtin-run.c b/builtin-run.c
+index 54e03be..40d9a17 100644
+--- a/builtin-run.c
++++ b/builtin-run.c
+@@ -633,18 +633,42 @@ static struct kvm *kvm_cmd_run_init(int argc, const char **argv)
+
+ static int kvm_cmd_run_work(struct kvm *kvm)
+ {
+-	int i;
+-	void *ret = NULL;
+-
+-	for (i = 0; i < kvm->nrcpus; i++) {
+-		if (pthread_create(&kvm->cpus[i]->thread, NULL, kvm_cpu_thread, kvm->cpus[i]) != 0)
+-			die("unable to create KVM VCPU thread");
+-	}
+-
+-	/* Only VCPU #0 is going to exit by itself when shutting down */
+-	return pthread_join(kvm->cpus[0]->thread, &ret);
++        int i;
++        int exit_status = 0;
++
++        for (i = 0; i < kvm->nrcpus; i++) {
++                if (pthread_create(&kvm->cpus[i]->thread, NULL, kvm_cpu_thread, kvm->cpus[i]) != 0)
++                        die("unable to create KVM VCPU thread");
++        }
++
++        /*
++         * Only VCPU #0 is going to exit by itself when shutting down
++         *
++         * BUT It's not actually true for reboot sequence
++         * Reboot is sending SIGKVMEXIT signal to all cpus
++         * and that is causing race condition on removing
++         * devices (which are pausing exiting cpus).
++         * This piece of code waits for death of all
++         * cpu threads
++         */
++        for (i = 0; i < kvm->nrcpus; i++) {
++                int c_thrstatus = 0, retval = 0;
++
++                void* thrstatus = &c_thrstatus;
++
++                retval = pthread_join(kvm->cpus[i]->thread, &thrstatus);
++
++                if (retval != 0)
++                        die("unable to end KVM VCPU thread");
++
++                /* Set exit status if one of vcpus returns error code > 0 */
++                if ((intptr_t) c_thrstatus != 0)
```

Comment:
pthread has confusing docs about the thread return value, so I wrote a small C example to remind myself how does it work exactly:
```c
#include <pthread.h>
#include <stdio.h>
#include <stdint.h>

void*
run(void* arg)
{
  return (void*)(intptr_t)1;
}

int
main(void)
{
  pthread_t tid;
  int retval = 0;
  void *ptr = &retval;
  pthread_create(&tid, NULL, run, NULL);
  pthread_join(tid, &ptr);
  printf("ptr: %p\nretval: %d\n", ptr, retval);
}
```

The result was:
```
ptr: 0x1
retval: 0
```

So, I think that:
1. It should be `(intptr_t) thrstatus` in both places (in the `if` and in its body).
2. The `c_thrstatus` variable is unnecessary and can be removed.
3. The `c_thrstatus` variable is used to initialize the value of the `thrstatus` variable. I guess we can skip the initialization altogether (or initialize it to something else like `&retval` or `NULL`, but it's not important - the value will be overwritten by `pthread_join` anyway).
4. Another problem I see here is that this patch changes semantics of the `kvm_cmd_run_work` function's return value - earlier it was returning an `errno` value (`EDEADLK`, `EINVAL`, and so on), now it returns a return value of thread's function and it is not a typical `errno` value, just `0` or `1`.